File: plugin/format-vue/performance-benchmark.js
/**
 * 🚀 Format-Vue 性能基准测试
 * 
 * 测试项目：
 * 1. 组件渲染性能
 * 2. 日志系统性能
 * 3. 文件列表渲染性能
 * 4. Watch响应性能
 * 5. 内存使用情况
 */

import { performance } from 'perf_hooks'
import { writeFileSync } from 'fs'

class PerformanceBenchmark {
  constructor() {
    this.results = []
    this.memoryBaseline = null
  }

  /**
   * 记录内存基线
   */
  recordMemoryBaseline() {
    if (global.gc) {
      global.gc()
    }
    this.memoryBaseline = process.memoryUsage()
    console.log('📊 Memory Baseline:', {
      heapUsed: `${(this.memoryBaseline.heapUsed / 1024 / 1024).toFixed(2)} MB`,
      heapTotal: `${(this.memoryBaseline.heapTotal / 1024 / 1024).toFixed(2)} MB`
    })
  }

  /**
   * 测试函数执行时间
   */
  async benchmark(name, fn, iterations = 1000) {
    console.log(`\n🔍 Testing: ${name}`)
    
    // 预热
    for (let i = 0; i < 10; i++) {
      await fn()
    }

    // 正式测试
    const times = []
    for (let i = 0; i < iterations; i++) {
      const start = performance.now()
      await fn()
      const end = performance.now()
      times.push(end - start)
    }

    // 统计
    const avg = times.reduce((a, b) => a + b, 0) / times.length
    const min = Math.min(...times)
    const max = Math.max(...times)
    const p95 = times.sort((a, b) => a - b)[Math.floor(times.length * 0.95)]

    const result = {
      name,
      iterations,
      avg: avg.toFixed(3),
      min: min.toFixed(3),
      max: max.toFixed(3),
      p95: p95.toFixed(3),
      unit: 'ms'
    }

    this.results.push(result)

    console.log(`  ✅ Avg: ${result.avg}ms | Min: ${result.min}ms | Max: ${result.max}ms | P95: ${result.p95}ms`)

    return result
  }

  /**
   * 测试内存使用
   */
  async benchmarkMemory(name, fn, iterations = 100) {
    console.log(`\n💾 Memory Test: ${name}`)

    if (global.gc) {
      global.gc()
    }

    const before = process.memoryUsage()

    for (let i = 0; i < iterations; i++) {
      await fn()
    }

    if (global.gc) {
      global.gc()
    }

    const after = process.memoryUsage()

    const heapDiff = (after.heapUsed - before.heapUsed) / 1024 / 1024
    const result = {
      name,
      iterations,
      heapDiff: heapDiff.toFixed(2),
      unit: 'MB'
    }

    this.results.push(result)

    console.log(`  ✅ Heap Diff: ${result.heapDiff} MB`)

    return result
  }

  /**
   * 生成报告
   */
  generateReport() {
    console.log('\n' + '='.repeat(80))
    console.log('📊 PERFORMANCE BENCHMARK REPORT')
    console.log('='.repeat(80))

    console.log('\n⏱️  Timing Results:')
    console.log('─'.repeat(80))
    this.results
      .filter(r => r.unit === 'ms')
      .forEach(r => {
        console.log(`${r.name.padEnd(40)} | Avg: ${r.avg}ms | P95: ${r.p95}ms`)
      })

    console.log('\n💾 Memory Results:')
    console.log('─'.repeat(80))
    this.results
      .filter(r => r.unit === 'MB')
      .forEach(r => {
        console.log(`${r.name.padEnd(40)} | Heap: ${r.heapDiff} MB`)
      })

    console.log('\n' + '='.repeat(80))

    return this.results
  }

  /**
   * 保存报告到文件
   */
  saveReport(filename = 'performance-report.json') {
    const report = {
      timestamp: new Date().toISOString(),
      baseline: this.memoryBaseline,
      results: this.results
    }
    writeFileSync(filename, JSON.stringify(report, null, 2))
    console.log(`\n💾 Report saved to: ${filename}`)
  }
}

// 模拟测试场景
async function runBenchmarks() {
  const bench = new PerformanceBenchmark()
  
  bench.recordMemoryBaseline()

  // 1. 测试日志添加性能
  await bench.benchmark('Log Entry Creation', () => {
    const now = new Date()
    const time = `${String(now.getHours()).padStart(2, '0')}:${String(now.getMinutes()).padStart(2, '0')}:${String(now.getSeconds()).padStart(2, '0')}`
    const log = {
      time,
      icon: '📝',
      message: 'Test log message',
      type: 'info'
    }
    return log
  }, 10000)

  // 2. 测试数组push性能
  const logs = []
  await bench.benchmark('Array Push (1000 items)', () => {
    logs.push({
      time: '12:00:00',
      icon: '📝',
      message: 'Test',
      type: 'info'
    })
  }, 1000)

  // 3. 测试对象深拷贝性能
  const obj = { a: 1, b: 2, c: { d: 3, e: 4 } }
  await bench.benchmark('Object Spread Copy', () => {
    return { ...obj }
  }, 10000)

  // 4. 测试setTimeout创建性能
  await bench.benchmark('setTimeout Creation', () => {
    return new Promise(resolve => {
      const timer = setTimeout(() => {
        clearTimeout(timer)
        resolve()
      }, 0)
    })
  }, 1000)

  // 5. 测试防抖setTimeout性能
  let timer = null
  await bench.benchmark('Debounced setTimeout', () => {
    if (timer) clearTimeout(timer)
    return new Promise(resolve => {
      timer = setTimeout(() => {
        timer = null
        resolve()
      }, 0)
    })
  }, 1000)

  // 6. 测试文件列表渲染数据准备
  await bench.benchmark('File List Data Prep (100 files)', () => {
    const files = []
    for (let i = 0; i < 100; i++) {
      files.push({
        id: `file-${i}`,
        name: `image-${i}.jpg`,
        ext: 'jpg',
        size: 1024 * 1024 * 2,
        thumbnail: null
      })
    }
    return files
  }, 100)

  // 7. 内存测试：大量日志
  await bench.benchmarkMemory('1000 Log Entries', () => {
    const logs = []
    for (let i = 0; i < 1000; i++) {
      logs.push({
        time: '12:00:00',
        icon: '📝',
        message: `Log message ${i}`,
        type: 'info'
      })
    }
    return logs
  }, 10)

  // 8. 内存测试：文件列表
  await bench.benchmarkMemory('1000 File Objects', () => {
    const files = []
    for (let i = 0; i < 1000; i++) {
      files.push({
        id: `file-${i}`,
        name: `image-${i}.jpg`,
        ext: 'jpg',
        size: 1024 * 1024 * 2,
        path: `/path/to/image-${i}.jpg`,
        thumbnail: null
      })
    }
    return files
  }, 10)

  bench.generateReport()
  bench.saveReport('plugin/format-vue/performance-report.json')
}

// 运行基准测试
console.log('🚀 Starting Format-Vue Performance Benchmark...\n')
runBenchmarks().catch(console.error)

export { PerformanceBenchmark }

File: ai-vue-refactor/core/scripts/performance_benchmark.py
#!/usr/bin/env python3
"""
🚀 Pixly Python ML 性能基准测试

运行方式：
python scripts/performance_benchmark.py

测试项目：
1. 特征提取性能
2. 模型预测性能
3. 数据预处理性能
4. 内存使用情况
5. 并发处理性能
"""

import time
import sys
import json
import numpy as np
from pathlib import Path
from typing import List, Dict, Any
import tracemalloc
import statistics

class PerformanceBenchmark:
    """性能基准测试类"""
    
    def __init__(self):
        self.results = []
        self.memory_baseline = None
    
    def record_memory_baseline(self):
        """记录内存基线"""
        tracemalloc.start()
        self.memory_baseline = tracemalloc.get_traced_memory()
        print(f"📊 Memory Baseline: {self.memory_baseline[0] / 1024 / 1024:.2f} MB")
    
    def benchmark(self, name: str, func, iterations: int = 1000):
        """测试函数执行时间"""
        print(f"\n🔍 Testing: {name}")
        
        # 预热
        for _ in range(10):
            func()
        
        # 正式测试
        times = []
        for _ in range(iterations):
            start = time.perf_counter()
            func()
            end = time.perf_counter()
            times.append((end - start) * 1000)  # 转换为毫秒
        
        # 统计
        avg = statistics.mean(times)
        min_time = min(times)
        max_time = max(times)
        p95 = statistics.quantiles(times, n=20)[18]  # 95th percentile
        
        result = {
            'name': name,
            'iterations': iterations,
            'avg': round(avg, 3),
            'min': round(min_time, 3),
            'max': round(max_time, 3),
            'p95': round(p95, 3),
            'unit': 'ms'
        }
        
        self.results.append(result)
        
        print(f"  ✅ Avg: {result['avg']}ms | Min: {result['min']}ms | Max: {result['max']}ms | P95: {result['p95']}ms")
        
        return result
    
    def benchmark_memory(self, name: str, func, iterations: int = 100):
        """测试内存使用"""
        print(f"\n💾 Memory Test: {name}")
        
        tracemalloc.start()
        before = tracemalloc.get_traced_memory()
        
        for _ in range(iterations):
            func()
        
        after = tracemalloc.get_traced_memory()
        tracemalloc.stop()
        
        heap_diff = (after[0] - before[0]) / 1024 / 1024
        
        result = {
            'name': name,
            'iterations': iterations,
            'heapDiff': round(heap_diff, 2),
            'unit': 'MB'
        }
        
        self.results.append(result)
        
        print(f"  ✅ Heap Diff: {result['heapDiff']} MB")
        
        return result
    
    def generate_report(self):
        """生成报告"""
        print('\n' + '=' * 80)
        print('📊 PERFORMANCE BENCHMARK REPORT')
        print('=' * 80)
        
        print('\n⏱️  Timing Results:')
        print('─' * 80)
        for r in self.results:
            if r['unit'] == 'ms':
                print(f"{r['name']:<40} | Avg: {r['avg']}ms | P95: {r['p95']}ms")
        
        print('\n💾 Memory Results:')
        print('─' * 80)
        for r in self.results:
            if r['unit'] == 'MB':
                print(f"{r['name']:<40} | Heap: {r['heapDiff']} MB")
        
        print('\n' + '=' * 80)
        
        return self.results
    
    def save_report(self, filename: str = 'performance-report-python.json'):
        """保存报告到文件"""
        report = {
            'timestamp': time.strftime('%Y-%m-%d %H:%M:%S'),
            'baseline': self.memory_baseline,
            'results': self.results
        }
        
        with open(filename, 'w') as f:
            json.dump(report, f, indent=2)
        
        print(f"\n💾 Report saved to: {filename}")


def run_benchmarks():
    """运行所有基准测试"""
    bench = PerformanceBenchmark()
    
    bench.record_memory_baseline()
    
    # 1. 测试NumPy数组操作
    bench.benchmark('NumPy Array Creation (128d)', lambda: np.random.rand(128), 10000)
    
    # 2. 测试特征归一化
    features = np.random.rand(128)
    bench.benchmark('Feature Normalization', lambda: features / np.sum(features), 10000)
    
    # 3. 测试矩阵乘法
    matrix_a = np.random.rand(128, 128)
    matrix_b = np.random.rand(128, 128)
    bench.benchmark('Matrix Multiplication (128x128)', lambda: np.dot(matrix_a, matrix_b), 1000)
    
    # 4. 测试列表操作
    bench.benchmark('List Comprehension (1000 items)', 
                   lambda: [i * 2 for i in range(1000)], 1000)
    
    # 5. 测试字典操作
    bench.benchmark('Dict Creation (100 items)', 
                   lambda: {f'key_{i}': i for i in range(100)}, 1000)
    
    # 6. 测试JSON序列化
    data = {'features': [0.5] * 128, 'quality': 90, 'speed': 4}
    bench.benchmark('JSON Serialization', lambda: json.dumps(data), 10000)
    
    # 7. 测试JSON反序列化
    json_str = json.dumps(data)
    bench.benchmark('JSON Deserialization', lambda: json.loads(json_str), 10000)
    
    # 8. 测试文件路径操作
    bench.benchmark('Path Operations', 
                   lambda: Path('/path/to/file.jpg').with_suffix('.jxl'), 10000)
    
    # 9. 内存测试：大数组
    bench.benchmark_memory('Large NumPy Array (1000x1000)', 
                          lambda: np.random.rand(1000, 1000), 10)
    
    # 10. 内存测试：特征向量列表
    bench.benchmark_memory('1000 Feature Vectors (128d)', 
                          lambda: [np.random.rand(128) for _ in range(1000)], 10)
    
    bench.generate_report()
    bench.save_report('scripts/performance-report-python.json')


if __name__ == '__main__':
    print('🚀 Starting Python ML Performance Benchmark...\n')
    try:
        run_benchmarks()
    except Exception as e:
        print(f"\n❌ Error: {e}")
        traceback.print_exc()
        sys.exit(1)
